### ISSUE TITLE
Neighbor list includes particles beyond cutoff

### ISSUE THREAD

**Issue by fabiannagel** [google/jax-md#149]
While trying to wrap my mind around how JAX-MD handles neighbor lists, I noticed that it includes particles beyond the cutoff radius. The following minimum working example reads an [ASE input geometry](https://github.com/google/jax-md/files/7164072/geometry.in.zip) in order to reproduce the problem.

```
from ase.io import read
import jax.numpy as jnp
import jax_md
from jax_md.partition import NeighborList
from jax_md.space import DisplacementFn


def get_input(geometry_file: str):
    atoms = read(geometry_file, format="aims")
    R = jnp.float32(atoms.positions)
    box = jnp.float32(atoms.get_cell().array)
    return R, box


def sanity_check_nl(R: jnp.ndarray, neighbor_list: NeighborList, displacement_fn: DisplacementFn,
                    r_cutoff: float, dr_threshold: float):
    
    for r, neighbor_indices in zip(R, neighbor_list.idx):
        neighbors_of_r = R[neighbor_indices]

        for neighbor_idx, neighbor in enumerate(neighbors_of_r):
            displacements = displacement_fn(r, neighbor)
            distance = jnp.sqrt(jnp.sum(displacements ** 2))

            if distance > (r_cutoff + dr_threshold):
                print("{} > {}".format(distance, r_cutoff + dr_threshold))
                

r_cutoff = 5.0
dr_threshold = 1.0
R, box = get_input("geometry.in")

displacement_fn, shift_fn = jax_md.space.periodic_general(box, fractional_coordinates=False)

neighbor_fn = jax_md.partition.neighbor_list(
    displacement_fn,
    box,
    r_cutoff,
    dr_threshold,
    capacity_multiplier=1.0,
    mask_self=False,
    disable_cell_list=True,
    fractional_coordinates=False)

neighbor_list = neighbor_fn(R)
sanity_check_nl(R, neighbor_list, displacement_fn, r_cutoff, dr_threshold)
```

**Comment by sschoenholz** [comment]
Thanks for the repro! Looking into this a bit more closely I think everything is working as intended. Basically I think the line, 

```python
for neighbor_idx, neighbor in enumerate(neighbors_of_r):
```

should read:

```python
for neighbor_idx, neighbor in zip(neighbor_indices, neighbors_of_r):
  if neighbor_idx == R.shape[0]:
    continue
```

Note that you can also check the distances by writing something like this,

```python
dist = space.distance(space.map_neighbor(displacement_fn)(R, R[neighbor_list.idx]))
print(jnp.max(dist * (neighbor_list.idx < R.shape[0])))
```

I will still need to dig into the fractional coordinate issue. Let me know whether there are any outstanding problems with this reasoning.

**Comment by fabiannagel** [comment]
Right, my bad! Obviously the neighbor indices are not equal to their enumeration indices. It's also good to know that `R.shape[0]` is used for padding. Thanks for your help!
